clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -237,12 +239,18 @@ public void close() {
      * @param send The request to send
      */
     public void send(Send send) {
-        KafkaChannel channel = channelOrFail(send.destination());
-        try {
-            channel.setSend(send);
-        } catch (CancelledKeyException e) {
-            this.failedSends.add(send.destination());
-            close(channel);
+        String connectionId = send.destination();
+        KafkaChannel closingChannel = closingChannels.remove(connectionId);
+        if (closingChannel != null)
+            doClose(closingChannel, true);
```

Comment:
I started with a close with three parameters exactly as you suggest. And then changed it because I thought it looked a bit confusing. The overload does make it clearer, so I have updated the code. Thank you.